template.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"fmt"
+	"math/rand"
+	"net/url"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig/v3"
+	"github.com/brianvoe/gofakeit"
+)
+
+type LogGenerator struct {
+	templateFuncMap   template.FuncMap
+	logFieldTemplates map[string]*template.Template
+}
+
+func NewLogGenerator() *LogGenerator {
+	templateFuncMap := sprig.TxtFuncMap()
+	templateFuncMap["randValue"] = func(values ...string) string {
+		return values[rand.Intn(len(values))]
+	}
+	templateFuncMap["randInt"] = func(min, max int) string {
+		return strconv.Itoa(gofakeit.Number(min, max))
+	}
+	templateFuncMap["randFloat"] = func(min, max float64) string {
+		return fmt.Sprintf("%.2f", gofakeit.Float64Range(min, max))
+	}
+	templateFuncMap["randIPAddress"] = func() string {
+		return gofakeit.IPv4Address()
+	}
+	templateFuncMap["randIPV4Address"] = func() string {
+		return gofakeit.IPv4Address()
+	}
+	templateFuncMap["randIPV6Address"] = func() string {
+		return gofakeit.IPv6Address()
+	}
+
+	return &LogGenerator{
+		templateFuncMap:   templateFuncMap,
+		logFieldTemplates: make(map[string]*template.Template),
+	}
+}
+
+func (g *LogGenerator) SetLogFieldTemplate(field, templateStr string) error {
+	tmpl, err := template.New(field).Funcs(g.templateFuncMap).Parse(templateStr)
+	if err != nil {
+		return err
+	}
+	g.logFieldTemplates[field] = tmpl
+	return nil
+}
+
+type TemplateData struct {
+	Hostname                    string
+	Priority                    string
+	Application                 string
+	PID                         string
+	Message                     string
+	NumberZeroToThousand        string
+	NumberZeroToTenThousand     string
+	NumberZeroToHundredThousand string
+	NumberZeroToMillion         string
+	IPAddress                   string
+	IPV4Address                 string
+	IPV6Address                 string
+	LocalIPAddress              string
+	LocalIPV4Address            string
+	LocalIPV6Address            string
+	RemoteIPAddress             string
+	RemoteIPV4Address           string
+	RemoteIPV6Address           string
+	Username                    string
+	HTTPMethod                  string
+	HTTPPath                    string
+	HTTPStatusCode              string
+	HTTPRequestSize             string
+	HTTPResponseSize            string
+	HTTPReferer                 string
+	HTTPAgent                   string
+	HTTPVersion                 string
+	HTTPProtocol                string
+	AuthUserID                  string
+	BuildVersion                string
+}
+
+func (g *LogGenerator) Generate() (map[string]string, error) {
+	var (
+		record = make(map[string]string, len(g.logFieldTemplates))
+		data   TemplateData
+		buffer strings.Builder
+	)
+
+	data = TemplateData{
```

Comment:
Why using template data? I think these could be template functions to give more control.
I think it makes more sense to use a function since its generating data, instead of accessing to some predefined values.

Reply:
Gofakeit already allows be be exposed to text/template https://github.com/brianvoe/gofakeit?tab=readme-ov-file#templates